js/drafts/composer.js

```diff
@@ -155,8 +155,15 @@ function Composer(newRecords, account, identity) {
 
   body.attachments.forEach(function(attachment) {
     try {
-      var attachmentNode = new MimeNode(attachment.type,
-                                        { filename: attachment.name });
+      var attachmentNode = new MimeNode(
+        attachment.type,
+        {
+          // This implies Content-Disposition: attachment
+          filename: attachment.name
+        });
+      // Explicitly specify base64 encoding; this is the default for non
+      // text/* parts, but it's important for us to not break on those, too.
+      attachmentNode.setHeader('Content-Transfer-Encoding', 'base64');
```

Comment:
So the bug that's being fixed here (bug 1111724) is that if the user attaches an attachment with any "text/" MIME type, we don't actually attach it.  Instead we get our magic-replacement marker that looks like "{{blob!0.56163309258718531418669397963}}" as the contents of the attachment.

This is caused because the code that does the replace assumes that the contents of the attachment were base64-encoded, and so it base64-encodes that marker.  This change forces us to encode all attachments as base64, even text/plain and friends.  mailbuild (the MIME building library we use for composition) does not do trade-off analysis between quoted-printable and base64 encodings at this time, although it does try and use 7bit if possible first.  And base64 is in the worst-case more efficient than quoted-printable, so this is arguably a sane choice to make.  (Not that you're debating it, just clarifying why just forcing base64 is a reasonable fix.)

I'll clarify this comment now.